### PR TITLE
Added java sdk

### PR DESCRIPTION
**Pull request by NgLoader**
## Summary
I have implemented a Java Software Development Kit (SDK) that supports logging and tracing.
This SDK is compatible with Java 17 and later versions.

The issue associated with this SDK implementation is #4364.

## Todo
Here are the tasks that need to be done:
- [x] Add documentation
- [x] Add JUnit tests
- [x] GitHub actions support

## How did you test this change?
I have only conducted local testing of this implementation, and haven't used any test classes so far.

## Are there any deployment considerations?
I wanted to let you know that deploying the implementation to a Maven repository is to make it more accessible to everyone.
I believe that this will greatly improve the accessibility of this implementation and help more people benefit from it.

If you have any questions or need further information about the deployment process, please feel free to contact me on Discord at ``Nils#0001``.
I will be happy to provide you with any assistance you need.

</br>
</br>

# About

### Severity
We have introduced a severity class that can determine the log level and provide additional details about it.
The severity class also includes a level priority, ranging from one to four.
More information about the severity fields can be found in [Opentelemetry's specification](https://opentelemetry.io/docs/reference/specification/logs/data-model/#severity-fields)

Example:
```java
Severity.INFO
Severity.info(Priority.NORMAL);
Severity.info("Core");
Severity.info("Core", Priority.HIGH);

```

### Usage
Currently, there are two constructors available for this SDK:

```java
Highlight.init("PROJECT_ID", options -> options
	.version("1.0.0")
	.environment("development")
	.attributes(attributes -> attributes
		.put("some.custom.attribute", true)
		.put("another.custom.attribute", "thats nice")));
```
or
```java
HighlightOptions options = HighlightOptions.builder("PROJECT_ID")
	.version("1.0.0")
	.environment("development")
	.attributes(attributes -> attributes
		.put("some.custom.attribute", true)
		.put("another.custom.attribute", "thats nice"))
	.build();

Highlight.init(options);
```

To log an exception, use the following line of code:
```java
Highlight.captureException(exception);
```

For everything else, use this code:
```java
Highlight.captureLog(Severity.INFO, "Just another message");
```

You can also apply more detailed logging or exceptions using the following code:
```java
Highlight.captureRecord(HighlightRecord.log()
	.severity(Severity.warn("Internal", Priority.HIGH))
	.message("Just another message")
	.requestId("REQUEST_ID")
	.attributes(attributes -> attributes
		.put("application.user.name", "NgLoader"))
	.build());
```

If you are using user sessions, you can create a session class that provides all the highlight methods, as shown below:
```java
HighlightSession session = new HighlightSession("SESSION_ID");
session.captureException(new NullPointerException("This shouldn't happen"));
session.captureLog(Severity.INFO, "Just another message");
session.captureRecord(HighlightRecord.log()
	.severity(Severity.warn("Internal", Priority.HIGH))
	.message("Just another message")
	.requestId("REQUEST_ID")
	.attributes(attributes -> attributes
		.put("application.user.name", "NgLoader"))
	.build());
```
Alternatively, you can add a session to a builder as shown below:
```java
// Existing session
HighlightSession session = new HighlightSession("SESSION_ID");
Highlight.captureRecord(HighlightRecord.error()
	.throwable(new NullPointerException("This shouldn't happen"))
	.userSession(session)
	.build());

// Without session
Highlight.captureRecord(HighlightRecord.error()
	.throwable(new NullPointerException("This shouldn't happen"))
	.userSession("SESSION_ID")
	.build());
```

### CODE REVIEW

**Comment by NgLoader** [comment]
Github Actions are now supported.
Currently I tested everything on [my fork](https://github.com/NgLoader/highlight/actions/workflows/java-sdk.yml) (please note that the failed builds are successful because there are currently unable to deploy)

**Comment by NgLoader** [comment]
Thanks

File: sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/Highlight.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+		this.openTelemetry = new HighlightOpenTelemetry(this);
+		this.tracer = new HighlightTracer(this);
+		this.logger = new HighlightLogger(this);
+
+		Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+			System.out.println("Highlight is shutting down...");
+			long startTime = System.currentTimeMillis();
+
+			this.openTelemetry.shutdown().join(10, TimeUnit.SECONDS);
+			// TODO Throw errors when try to log after shutdown
```
Comment (Vadman97) [inline sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/Highlight.java:80]: that would be a good addition! i'd assume that the `openTelemetry.shutdown()` method is atomic though (if a new trace is submitted while the shutdown is happening, then opentelemetry would refuse it or flush it) and if the `.join` times out, that will raise an exception

File: sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/Highlight.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+			long finishedTime = System.currentTimeMillis() - startTime;
+			System.out.println("Highlight was successfully shutted down in " + finishedTime + "ms.");
```
Comment (Vadman97) [inline sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/Highlight.java:83]:
```suggestion
			System.out.println("Highlight was successfully shut down in " + finishedTime + "ms.");
```

File: sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/HighlightLogger.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	private final String projectId;
+
+	private final Logger logger;
+
+	HighlightLogger(Highlight highlight) {
+		HighlightOptions options = highlight.getOptions();
+		this.projectId = options.projectId();
+
+		HighlightOpenTelemetry openTelemetry = highlight.getOpenTelemetry();
+		this.logger = openTelemetry.getLogger("highlight-node");
```
Comment (Vadman97) [inline sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/HighlightLogger.java:21]:
```suggestion
		this.logger = openTelemetry.getLogger("highlight-java");
```

File: sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/HighlightOpenTelemetry.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+	private final ContextPropagators propagator;
+
+	HighlightOpenTelemetry(Highlight highlight) {
+		HighlightOptions options = highlight.getOptions();
+
+		AttributesBuilder attributesBuilder = Attributes.builder()
+				.put(HighlightAttributes.HIGHLIGHT_PROJECT_ID, options.projectId())
+				.put(ResourceAttributes.TELEMETRY_SDK_LANGUAGE, "java") // TODO compile with maven
+				.put(ResourceAttributes.TELEMETRY_SDK_NAME, "highlight") // TODO compile with maven
```
Comment (Vadman97) [inline sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/HighlightOpenTelemetry.java:43]: what does `compile with maven` mean? do you mean the `highlight` constant can come from maven?

File: sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/HighlightOpenTelemetry.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+				.setResource(resource)
+				.build();
+
+		// Propagator
+		TextMapPropagator textMapPropagator = TextMapPropagator.composite(
+				B3Propagator.injectingMultiHeaders(),
+                W3CTraceContextPropagator.getInstance());
+
+		this.propagator = ContextPropagators.create(textMapPropagator);
+	}
```
Comment (Vadman97) [inline sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/HighlightOpenTelemetry.java:84]: what does the propagator do?

**Review by Vadman97 — APPROVED** [review]
amazing, thank you for getting this in @NgLoader ! 
do you mind adding some docs on using this SDK? for example, [here are docs](https://github.com/highlight/highlight/blob/main/highlight.io/components/QuickstartContent/backend/go/go-gqlgen.tsx#L4) for adding gqlgen instrumentation. could you add something similar for java?
more [docs on how to contribute to the docs site here](https://www.highlight.io/docs/general/company/open-source/contributing/landing-site).

File: sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/HighlightOpenTelemetry.java (hunk 4 above)
Comment (NgLoader) [reply sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/HighlightOpenTelemetry.java:43]: Maven can automaticly insert the currently build version from there pom.xml

File: sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/Highlight.java (hunk 1 above)
Comment (NgLoader) [reply sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/Highlight.java:80]: Yea, but i think when the error comes from highlight it makes more sense

File: sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/HighlightOpenTelemetry.java (hunk 5 above)
Comment (NgLoader) [reply sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/HighlightOpenTelemetry.java:84]: It is used for the tracing and logging request you can read more about that [here](https://opentelemetry.io/docs/reference/specification/context/api-propagators/)

Comment (NgLoader) [reply sdk/highlight-java/highlight-sdk/src/main/java/io/highlight/sdk/HighlightOpenTelemetry.java:84]: Maybe I need to check the composits again

**Comment by NgLoader** [comment]
> amazing, thank you for getting this in @NgLoader ! 
> do you mind adding some docs on using this SDK? for example, [here are docs](https://github.com/highlight/highlight/blob/main/highlight.io/components/QuickstartContent/backend/go/go-gqlgen.tsx#L4) for adding gqlgen instrumentation. could you add something similar for java?
> more [docs on how to contribute to the docs site here](https://www.highlight.io/docs/general/company/open-source/contributing/landing-site).

Yea, I can do that later

**Comment by doucheng** [comment]
Great ~

**Review by Vadman97 — APPROVED** [review]

**Comment by NgLoader** [comment]
Thanks @Vadman97

**Review by Vadman97 — APPROVED** [review]

**Review by Vadman97 — APPROVED** [review]
awesome job! what would be useful next is an example (and any necessary SDK helpers) for plugging this into a common java backend web framework like `Spring` to automatically capture errors and logs. would be great to see a quickstart doc added as well - https://github.com/highlight/highlight/blob/main/highlight.io/components/QuickstartContent/QuickstartContent.tsx

**Comment by NgLoader** [comment]
> awesome job! what would be useful next is an example (and any necessary SDK helpers) for plugging this into a common java backend web framework like `Spring` to automatically capture errors and logs. would be great to see a quickstart doc added as well - https://github.com/highlight/highlight/blob/main/highlight.io/components/QuickstartContent/QuickstartContent.tsx

okay

**Pull request merged**